### PR TITLE
Improve non-tuned param ordering

### PR DESCRIPTION
**Pull request by manupak**
This commit improves the init param ordering
to be sorted based on padding amount.
i.e. lowest padding is preffered.

When in equal padding cases, the original
ordering will be respected.

Doing so, exposed some params should be
dynamically filtered out according to
XdlopsCodeSelection.

### CODE REVIEW

**Comment by manupak** [comment]
Putting this to draft as this build upon https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/757 and blocked on merging the above.

**Comment by manupak** [comment]
For the exclusive changes in this PR, please look at 355126df37bd98a10229fc64b33c50524a7a49fd

**Comment by krzysz00** [comment]
Worth mentioning, you can set the base branch to be the other PR - that'll work fine

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h]: `const std::vector<T> &` is spelled `ArrayRef<T>`

File: mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h]: Nice catches

File: mlir/lib/Dialect/MIOpen/Tuning/GridwiseGemmParams.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/MIOpen/Tuning/GridwiseGemmParams.cpp]: That doesn't look right - I think you want padM to be 0 in the non-padding case. You probably want `mod_1_to_n` from the math utilities header here

Comment (manupak) [reply mlir/lib/Dialect/MIOpen/Tuning/GridwiseGemmParams.cpp]: Yes, you are right -- reused the existing padding calculating utils there and refactored it a bit.

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h
Comment (manupak) [reply mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h]: Done

**Comment by manupak** [comment]
> Worth mentioning, you can set the base branch to be the other PR - that'll work fine

Other PR branch is in my fork right ? this is a missing feature request for GH for a while now :sigh.
https://github.com/orgs/community/discussions/4477#discussioncomment-1039627

**Review by krzysz00 — COMMENTED** [review]
Code-wise, looks excellent!

Can we get perf comparisons before merging?

**Comment by manupak** [comment]
Sure, How would you generally do that (other than running a random assortment)? Any pointers are appreciated

**Comment by krzysz00** [comment]
The way I usually do it is to run (after doing a `ninja check-rocmlir-build-only ci-performance-scripts`):

```
./bin/MIOpenDriver.py -b -o=some-name.csv
echo "description" >perf-run-date
[cd to a checkout of upstream]
ninja check-rocmlir-build-only ci-performance-scripts
./bin/MIOpenDriver.py -b -o=upstream.csv
echo "Upstream" >perf-run-date
[cd back]
python3 ./bin/perfRegressionReport.py [path-to-upstream].csv [path-to-here].csv output-file.html
```
then copy the file off, take a look at it, and post it here

**Comment by krzysz00** [comment]
(You can, alternatively, move the results from your perf test along with the perf-run-date to another directory, then `git checkout miopen-dialect`, recompile, and run the script again)

**Comment by manupak** [comment]
Noted with thanks! I ll have a go tomorrow

**Comment by manupak** [comment]
https://gist.github.com/manupak/a1e679b89b1e05fe3db0a74b4879a266

Overall there is an increase there are few cases it drops..
It is worth noting the differences are in cases where they are believed to be generating a padded gemm.

**Comment by manupak** [comment]
ok some worse numbers does not seem to reproduce now and running multiple measurements to see...

**Comment by manupak** [comment]
I did three runs for the PR : 

1.) https://gist.github.com/manupak/887bcf5a3727f22ccfacff85b20603ac
2.) https://gist.github.com/manupak/1dcf61e381f472100f443bb2f8a52288
3.) https://gist.github.com/manupak/4156c2e8b9d09228d744a71ac1aee8b2

Generally looking at the changes only a single worse config stood out -- others more or less seemed like glitches.

| Direction | DataType | Layout | C |  H  |  W  |  K | Y | X | StrideH | StrideW | PaddingH | PaddingW | Unnamed: 0_new | MLIR TFlops (new) | Unnamed: 0_old | MLIR TFlops (old) | Current/Previous |
|:---------:|:--------:|:------:|:-:|:---:|:---:|:--:|:-:|:-:|:-------:|:-------:|:--------:|:--------:|:--------------:|:-----------------:|:--------------:|:-----------------:|:----------------:|
| fwd       | f16      | nhwc   | 3 | 230 | 230 | 64 | 7 | 7 |    2    |    2    |     0    |     0    |       68       |       11.75       |       68       |       21.11       |       0.56       |

It is interesting that it is also layout specific as the nchw layout seemed fine with config.

**Comment by krzysz00** [comment]
Having looked, those are certainly substantial wins, and the fact that we now have problems with one fp16 config isn't anywhere near that big a deal in my book.

So congrats on your perf win!

**Review by krzysz00 — APPROVED** [review]

**Comment by manupak** [comment]
Thanks for reviews @krzysz00  and working through with me with this.

**Pull request merged**
